definition_validator/variance.cc

```diff
@@ -0,0 +1,184 @@
+
+#include "common/typecase.h"
+#include "core/Symbols.h"
+#include "core/core.h"
+#include "core/errors/resolver.h"
+
+#include "variance.h"
+
+using namespace std;
+
+namespace sorbet::definition_validator::variance {
+
+class VarianceValidator {
+private:
+    const core::Loc loc;
+
+    VarianceValidator(const core::Loc loc) : loc(loc) {}
+
+    core::Variance invertVariance(const core::Variance variance) {
+        switch (variance) {
+            case core::Variance::CoVariant:
+                return core::Variance::ContraVariant;
+            case core::Variance::Invariant:
+                return core::Variance::Invariant;
+            case core::Variance::ContraVariant:
+                return core::Variance::CoVariant;
+        }
+    }
+
+    string showVariance(const core::Variance variance) {
+        switch (variance) {
+            case core::Variance::CoVariant:
+                return ":out";
+            case core::Variance::Invariant:
+                return ":invariant";
+            case core::Variance::ContraVariant:
+                return ":in";
+        }
+    }
+
+    bool compatibleVariance(const core::Variance left, const core::Variance right) {
+        switch (left) {
+            case core::Variance::CoVariant:
+                return right != core::Variance::ContraVariant;
```

Comment:
Why is `CoVariant` compatible with `Invariant`? (and the same for `ContraVariant`) Maybe an example would help.

Reply:
this can be implemented with `polarity * variance < 0`

Reply:
It's naming & terminology introduced in this PR that confused you. 
Having an invariant type member in contravariant position is fine, this is why it's `compatible`.

Reply:
I've restructured this to check for compatibility between the current polarity and the variance of the parameter. Let me know if this clears up the intent.